cachelib/allocator/CacheItem.h

```diff
@@ -305,12 +305,17 @@ class CACHELIB_PACKED_ATTR CacheItem {
    */
   RefcountWithFlags::Value getRefCountAndFlagsRaw() const noexcept;
 
-  FOLLY_ALWAYS_INLINE void incRef() {
-    if (LIKELY(ref_.incRef())) {
-      return;
+  // Increments item's ref count
+  //
+  // @return true on success, failure if item is marked as exclusive
+  // @throw exception::RefcountOverflow on ref count overflow
+  FOLLY_ALWAYS_INLINE bool incRef() {
```

Comment:
Not a fan of changing `incRef` behavior to have 3 outcomes (success, fail due to exclusivity, refcount overflow). It complicates existing behavior.

My bigger concern is actually that I don't see a clear need for introducing `markedForEviction`.

Today the goal of preventing anyone else from accessing an item while it is being prepared for eviction is to remove it from accessContainer. Why do we need this change?

Reply:
> @igchor: can you help me understand why we need this change?
> 
> Today the goal of preventing anyone else from accessing an item while it is being prepared for eviction is to remove it from accessContainer. Is this insufficient?
> 
> My general impression reading the code is that we're changing the responsibility of managing an Item's state (and the right to access an item) to be exclusively within CacheItem logic itself. I don't understand why we're making this change. The change itself also doesn't seem to have reduced the complexity of eviction/moving management.

The main goal of this change is to improve performance.  This is a dependency of https://github.com/facebook/CacheLib/pull/172

As you noticed, marking an item for eviction is pretty much equivalent to removing it from `accessContainer`. However, the key difference is that `markForEviction` is lock-free. This is important because we would like to shorten MMContainer critical section (and possibly use combined locking), as I mentioned on https://github.com/facebook/CacheLib/pull/172

We could just call `markMoving`  instead of `markForEviction` inside MMContainer Critical Section. However, since this doesn't prevent users from obtaining new handles, the item marked as moving is not guaranteed to be evicted. This exact problem was causing a lower eviction success rate with my initial Critical Section Patch (which you found when running throughput/small_cache_chained_items.json)

Reply:
Another reason is that we would like to leverage `moving` bit for data movement between memory tiers. The idea is to block readers (create a wait context) for items marked as `moving`. 

This means that we cannot call `markMoving` inside `findEviction` for items which are being evicted to NVM/deleted.  We need to have a different mechanism to synchronize evictions and SlabRelease: i.e. `markForEviction`

Reply:
@therealgymmy here's a pseudocode of how we implemented support for multi-tier using `markMoving` and `markForEviction`:

```c++
//////////////////////////////////////////////////////////////////////////////
findEviction() {
	Item* candidate = NULL;
	mmContainer.withEvictionIterator(auto itr) {
		while (itr) {
			auto candidateMarked = 
				(lastTier && itr.get()->markForEviction()) ||
				(!lastTier && itr.get()->markMoving());
			if (candidateMarked) {
				// at this point, no one can obtain handle to the 'candidate'
				// if the item was marked for eviction, all readers will get NULL
				// if the item was marked moving, all readers will block until
				// wakeupReaders() is called
				candidate = itr.get();
				return;
			}
			
			itr++;
		}
	};
	
	if (!candidate) return NULL;
	
	if (!lastTier) {
		// tryEvictToNextMemoryTier can incur significant overhead - it might try to
		// evict the item on the layer below to make room for 'candidate'
		auto newHandle = tryEvictToNextMemoryTier(candidate);
		
		// item was successfuly moved between tiers
		if (newHandle) {
			// wakes up all readers waiting for the item
			wakeupReaders(candidate, newHandle);
		} else {
			// we failed to move the item to different tier, time to evict
			candidate->markExlusiveWhenMoving();
			removeFromAccessContainerAndevictItemToNVM(candidate);
			
			wakeupReaders(candidate, NULL);
			
		}
	} else {
		// since this is the last tier, candidate has already been marked for eviction,
		// evicition, will always suceed
		removeFromAccessContainerAndevictItemToNVM(candidate);
	}
	
	return candidate;
}

//////////////////////////////////////////////////////////////////////////////
// acquire is called when getting handle from AccessContainer (e.g. on find())
acquire(Item *it) {
	auto incRes = incRef(*it);
	if (incRes == incOK) {
		return WriteHandle(it);
	} else if (incRes == incFailedMarkedForEviction) {
		// item is being evicted
		return NULL;
	} else { // incRes = incFailedMoving
		// create wait context and wait until item is no longer moving
		return handleWithWaitContextForMovingItem(*it);
	}
}
```

Reply:
@igchor: thank you for clarifying with the pseudo-code. This now makes sense to me.

Re: markForEviction()

Can we set this bit to true even if there're concurrent accesses? The idea is to block future readers, and we can move this item to another queue for async eviction. (Obviously if 0 access, we can evict synchronously)

Not a suggestion for this diff but I am wondering if this will make the aggregate cost of evictions go down. Fewer skips are needed overtime.

Reply:
That's a good idea. I think we have considered something like this in the past and even created some implementation but we didn't have a way to block readers back then - we just put the items from the head/tail of the list into the queue and relied on BG workers to move the items. Now, once we can block the readers, we should probably revisit this.

Reply:
However, I think this will only work with `markMoving` and not `markForEviction` since the implementation uses `refCount` to distinguish between `forEviction` and `moving` states.

Reply:
@igchor  Do we want to change `acquire` now? After this PR merges, when the item has exclusive bit set, `acquire` will create a handle without incrementing refcount.

Reply:
Yes, you're right, it's better to change it now to have consistent behavior. Done.

Reply:
In this PR when findEviction, we mark an item as moving (exclusive).
Then in `advanceIteratorAndTryEvictRegularItem` we call accessContainer_.removeIf, it will fail because it fails to increment the refcount (exclusive). It doesn't seem right?

Reply:
Right now only `markForEviction` causes `incRef` to return NULL handle. Users can still obtain handles to items that are `markedMoving` (and we are using `markMoving` in advanceIteratorAndTryEvictRegularItem).

In the next step we want to change the behavior of `advanceIteratorAndTryEvictRegularItem` to actually rely on `markForEviction` but this will require some additional changes so I have not included them here. You can find them on https://github.com/facebook/CacheLib/pull/172

Reply:
Got it now. Thanks for explaining.